Retain asset without data for RENDER_WORLD-only assets

robtfm opened the pull request:
# Objective

when `RenderAssets` with `RenderAssetUsages::RENDER_WORLD` and without `RenderAssetUsages::MAIN_WORLD` are extracted, the asset is removed from the assets collection. this causes some issues:
- systems which rely on the asset, like picking with meshes, fail with "asset not found" errors which are unintuitive.
- loading the asset by path a second time results in the asset being reloaded from storage, re-extracted and re-transferred to gpu, replacing the existing asset
- knowledge about the asset state is lost, we cannot tell if an asset is already loaded with `AssetServer::get_handle`
- metadata (image size, e.g.) is no longer available for the asset

## Solution

### extraction:
- add `take_gpu_data` to the `RenderAsset` trait. use it to pull the data out of the asset for transfer, and leave the empty asset in the collection. default implementation just `clone`s the asset.
- if the data has already been taken, ~~panic. this follows from modifying an asset after extraction, which is always a code error, so i think panic here makes sense~~ _log an error_

### Mesh/RenderMesh:
- make `Mesh::attributes` and `Mesh::indices` options
- take them on extraction
- `expect` operations which access or modify the vertex data or indices if it has been extracted. accessing the vertex data after extraction is always a code error. fixes #19737 by resulting in the error `Mesh has been extracted to RenderWorld. To access vertex attributes, the mesh must have RenderAssetUsages::MAIN_WORLD`
- provide `try_xxx` operations which allow users to handle the access error gracefully if required (no usages as part of this pr, but provided for future)
- compute the mesh `Aabb` when gpu data is taken and store the result. this allows extracted meshes to still use frustum culling (otherwise using multiple copies of an extracted mesh now panics as `compute_aabb` relied on vertex positions). there's a bit of a tradeoff here: users may not need the Aabb and we needlessly compute it. but i think users almost always do want them, and computing once (for extracted meshes) is cheaper than the alternative, keeping position data and computing a fresh `Aabb` every time the mesh is used on a new entity.

### Image/GpuImage:
images are a little more complex because the data can be deliberately `None` for render-targets / GPU-written textures where we only want an uninitialized gpu-side texture.
- take `Image::data` on extraction
- record on the resulting `GpuImage` whether any data was found initially
- on subsequent modifications with no data, panic if there was data previously

corner case / issue: when used with `RenderAssetBytesPerFrameLimiter` there may be no previous gpu asset if it is still queued pending upload due to the bandwidth limit. this can result in a modified image with initial data skipping the `had_data` check, resulting in a blank texture. i think this is sufficiently rare that it's not a real problem, users would still hit the panic if the asset is transferred in time and the problem/solution should be clear when they do hit it.

### ShaderStorageBuffer/GpuShaderStorageBuffer
follows the same pattern as Image/GpuImage:
- take `ShaderStorageBuffer::data` on extraction
- record on the resulting `GpuShaderStorageBuffer` whether any data was found initially
- on modifications with no data, panic if there was data previously

we don't have the queue issue here because `GpuShaderStorageBuffer` doesn't implement `byte_len` so we can't end up queueing them.

#### other RenderAssets
i didn't modify the other `RenderAsset` types (`GpuAutoExposureCompensationCurve`, `GpuLineGizmo`, `RenderWireframeMaterial`, `PreparedMaterial`, `PreparedMaterial2d`, `PreparedUiMaterial`) on the assumption that ~~cloning these is cheap enough anyway~~ _the asset usages are not exposed so we should never call `take_gpu_data`. the default implementation panics with a message directing users to implement the method if required_

## Testing

only really tested within my work project. i can add some explicit tests if required.

jasmine-nominal commented on crates/bevy_asset/src/render_asset.rs:
```diff
@@ -4,12 +4,13 @@ use serde::{Deserialize, Serialize};
-    /// If an asset is set to the `RENDER_WORLD` but not the `MAIN_WORLD`, the asset will be
-    /// unloaded from the asset server once it's been extracted and prepared in the render world.
+    /// If an asset is set to the `RENDER_WORLD` but not the `MAIN_WORLD`, the asset data (pixel data,
+    /// mersh vertex data, etc) will be removed from the cpu-side asset once it's been extracted and prepared
```
```suggestion
    /// mesh vertex data, etc) will be removed from the cpu-side asset once it's been extracted and prepared
```

andriyDev commented on crates/bevy_render/src/render_asset.rs:
```diff
@@ -86,6 +86,18 @@ pub trait RenderAsset: Send + Sync + 'static + Sized {
+    fn take_gpu_data(
+        source: &mut Self::SourceAsset,
+        _previous_gpu_asset: Option<&Self>,
+    ) -> Option<Self::SourceAsset> {
```
I'm not a fan of this Option. Can we make the return type `Result<Self::SourceAsset, InvalidModification>` or something? None isn't clear that it indicates a problem.

andriyDev commented on crates/bevy_mesh/src/mesh.rs:
```diff
@@ -389,19 +412,21 @@ impl Mesh {
 
     /// Retrieves the vertex `indices` of the mesh mutably.
     #[inline]
     pub fn indices_mut(&mut self) -> Option<&mut Indices> {
```
All these mutation methods essentially stop functioning when the asset is extracted. What about runtime generated (and mutated) assets? Is the intent that users will like fully replace the Mesh with like `*mesh_mut = Mesh::new()`?

If so, this probably needs a migration guide.

robtfm replied:
the pattern would be either to fully initialize in one step, or to initialize with `RenderAssetUsages::MAIN_WORLD`, make any modifications, then when it's finished change the `usage to RenderAssetUsages::RENDER_WORLD`. after that no more modifications are possible since the data is only on GPU and we can't get it back. 

if users want to make continuous runtime modifications, then they should use `MAIN_WORLD || RENDER_WORLD`, or fully replacing will also work.

this PR doesn't change anything in those scenarios, modifying an asset with `RENDER_WORLD` and not `MAIN_WORLD` is not possible currently either since the asset gets removed from the `Assets` collection. i can write something if you like, but the usage doesn't change.

andriyDev commented on crates/bevy_render/src/texture/gpu_image.rs:
```diff
@@ -35,6 +36,20 @@ impl RenderAsset for GpuImage {
+    ) -> Option<Self::SourceAsset> {
+        let data = source.data.take();
+
+        let valid_upload = data.is_some() || previous_gpu_asset.is_none_or(|prev| !prev.had_data);
```
Why can't we just reupload here? We're losing a capability here I think, since previously hot-reloading an image would just send the data to the GPU, which would unconditionally upload. Am I missing something here?

Could you also test that hot reloading still works? I believe the `alter_sprite` example is probably sufficient with the `file_watcher` feature (and then messing with the loaded sprite).

robtfm replied:
so the purpose of `RenderAssetUsages::RENDER_WORLD` is to reduce CPU memory by not keeping data on the CPU that is only required on the GPU. we clearly can't re-upload data from the CPU that we have chosen to purge.

as i described in the top comment, this check is specifically to catch an error which was not previously possible: where a user modifies some metadata on a `RENDER_WORLD`-only image after it has been transferred. that's not possible currently because currently the image gets removed from the CPU-side `Assets` collection so cannot be accidentally modified.

keeping access to it CPU side has the benefits i listed right at the top, but introduces this new way to break things, so i added this check to help anyone who does so accidentally. without this check, the image would be treated as an uninitialized texture -- for images with `data: None` we create an uninitialized GPU-texture which is much faster -- and the GPU pixel data would be zeros, which will never be what the user wants if the image had data when first transferred.

we're not losing any capabilities. default is `MAIN_WORLD | RENDER_WORLD`, in which case you can modify and (re)upload as much as you like because we keep the pixel data CPU side.

hot reloading still works for `MAIN_WORLD | RENDER_WORLD` and for `RENDER_WORLD`-only as reloading an asset reloads the data too.

andriyDev commented on crates/bevy_render/src/render_asset.rs:
```diff
@@ -86,6 +86,18 @@ pub trait RenderAsset: Send + Sync + 'static + Sized {
+    fn take_gpu_data(
+        source: &mut Self::SourceAsset,
+        _previous_gpu_asset: Option<&Self>,
+    ) -> Option<Self::SourceAsset> {
```
Yes, require users to implement the trait method. If they really are willing to tolerate cloning all their data, I think that's fine, but that should be an explicit choice from the impl.

andriyDev replied:
I much prefer making users implement this explicitly - pretty much every render asset should want to handle this correctly I think, especially since the default is just breaking.

robtfm commented on crates/bevy_mesh/src/mesh.rs:
```diff
@@ -389,19 +412,21 @@ impl Mesh {
     #[inline]
     pub fn indices(&self) -> Option<&Indices> {
-        self.indices.as_ref()
+        self.indices.as_ref().expect("Mesh has been extracted to RenderWorld. To access vertex attributes, the mesh must have RenderAssetUsages::MAIN_WORLD").as_ref()
```
making these return results will add literally hundreds of unwraps to the code base, e.g. there are 104 from primitive mesh builder functions calling `with_inserted_attribute` which are all guaranteed to succeed since they operate on meshes they've just constructed themselves.

perhaps i can add try_xxx variants, so that if we find situations where there are panics that we don't want we have a way to manage them in future.

robtfm commented:
- added MeshExtractableData enum to avoid the nasty nested options / make the intent clearer.
- added try_xxx functions to avoid 100s of unwraps in the cases where we know the access is valid (and reduce the breaking changes to the public api).
- changed the extract functions to return an error, and the default impl to return `AssetExtractionError::NoExtractionImplementation`.

i still need to do a pass through to fix up docs for all the copy/pasted functions, but the logic should be complete now i think.

greeble-dev commented on crates/bevy_mesh/src/mesh.rs:
```diff
@@ -1362,23 +2135,102 @@ impl Mesh {
+            values: VertexAttributeValues::Float32x3(position_values),
+            ..
+        }) = attributes.as_ref()?.get(&Self::ATTRIBUTE_POSITION.id)
+        {
```
```suggestion
        }) = attributes.as_ref()?.get(&Self::ATTRIBUTE_POSITION.id)
            && !position_values.is_empty()     
        {
```

Not 100% sure it's actually possible to make and extract a mesh with an empty position attribute, but seems reasonable to avoid any chance of panicking.

greeble-dev reviewed (changes requested):
Code seems reasonable - just got a couple of requests to avoid some corner cases. I also tested a variety of examples, and hacked in a few extra tests (including glTFs with `RENDER_WORLD` only, and making some examples deliberately trigger errors).

I don't like how the `Mesh` interface has ended up, but I'm not sure that's a problem with this PR. My hot take is that it's a problem with `Mesh` trying to be three things at once:

1. A finished, immutable asset that can be efficiently shared and rendered.
2. A way to incrementally build a mesh by adding/modifying attributes and suchlike.
3. A way to animate meshes by mutating an asset in-place and re-extracting.

I think the solution could be:

- Add a new `MeshBuilder` struct and move all the `Mesh` mutation methods to `MeshBuilder`.
  - This struct wouldn't have to deal with `MeshAccess` errors since it's entirely CPU side, so a load of `try_xxx` methods disappear.
  - A `finish` function would consume the builder and return a `Mesh`.
- Change `Mesh` to be immutable, except by `take_gpu_data`.
  - Animation that wants to mutate the mesh now has to create a fresh `Mesh` and overwrite the entire asset.

That gives as clear separation between 1) and 2), but makes 3) more awkward and inefficient. I'd argue that's ok in the short-term since it still basically works and mutating the asset is always going to be somewhat inefficient. In the long-term there could be a separate way of efficiently sending fine-grained mesh updates to the GPU.

The `Mesh`/`MeshBuilder` separation might also make compressed and derived data (like AABBs) more robust and efficient. So `MeshBuilder::finish` would automatically calculate AABBs and optionally apply compression.

EDIT: I should have said that I don't think this PR needs to explore stuff like `MeshBuilder`. What the PR does right now - make the `Mesh` interface more awkward in the short-term - seems like a reasonable trade-off for solving `RENDER_WORLD` issues now.

robtfm commented:
it would be nice to have a MeshBuilder with infallible access methods. then all the Mesh accessors could be result-based without scattering 100s of unwraps about.

i don't think i want to do it as part of this PR though.

greeble-dev reviewed (changes requested):
Noticed one more issue while testing meshes with missing attributes.

robtfm commented on crates/bevy_mesh/src/mesh.rs:
```diff
@@ -150,6 +254,9 @@ pub struct Mesh {
     pub enable_raytracing: bool,
+    /// Precomputed min and max extents of the mesh position data. Used mainly for constructing `Aabb`s for frustum culling.
+    /// This data will be set if/when a mesh is extracted to the GPU
+    pub final_aabb_extents: Option<(Vec3, Vec3)>,
```
changed the type, thanks.

it's used in bevy_camera so can't be private unless we add an accessor ... that might be better since it would stop users setting it wrongly, but would prevent users who have a legitimate reason to modify it ... not sure if there is such a reason.

beicause commented:
Regarding the Mesh interface, I‘d prefer to remove these `try_xxx` methods and let them panic. Adding a `is_extracted_to_render_world` method for checking is sufficient IMO, these `try_xxx` methods are rarely useful.

robtfm commented:
> Regarding the Mesh interface, I‘d prefer to remove these `try_xxx` methods and let them panic. Adding a `is_extracted_to_render_world` method for checking is sufficient IMO, these `try_xxx` methods are rarely useful.

since the change to results was at request of @andriyDev (and @alice-i-cecile) i'll defer to them. i agree fwiw

beicause commented on crates/bevy_mesh/src/mesh.rs:
```diff
@@ -128,12 +232,12 @@ pub struct Mesh {
+    morph_targets: MeshExtractableData<Handle<Image>>,
     #[cfg(feature = "morph")]
-    morph_target_names: Option<Vec<String>>,
+    morph_target_names: MeshExtractableData<Vec<String>>,
```
These extractable data can be consolidated. This ensures you always extract it together, and that they are either all with data or without data  simultaneously. Like this:
```rs
struct MeshExtractableData {
    attributes: BTreeMap<MeshVertexAttributeId, MeshAttributeData>,
    indices: Option<Indices>,
    #[cfg(feature = "morph")]
    morph_targets: Option<Handle<Image>>,
    #[cfg(feature = "morph")]
    morph_target_names: Option<Vec<String>>,
}

struct Mesh {
    ...
    extractable_data: Option<MeshExtractableData>,
    ...
}
```

robtfm replied:
could be. i don't think it improves anything but i can change it if others disagree.

IceSentry reviewed (approved):
The new `_option` function feel a bit weird but I don't really have an alternative solution. I mean it could either always return the option or always map the option to a result by it's not a blocker.

LGTM

beicause commented:
Another possible approach I consider is to consolidate the extractable data and access it through something similar to entry API, thereby avoiding multiple unwrap operations.
